app/data_client.go

```diff
@@ -177,13 +187,111 @@ type DatabaseConnReturn struct {
 	HasDatabaseUser bool
 }
 
+// DataSyncClient structs
+
+// SensorMetadata contains the time the sensor data was requested and was received.
+type SensorMetadata struct {
+	TimeRequested time.Time
+	TimeReceived  time.Time
+	MimeType      MimeType
+	Annotations   Annotations
+}
+
+// SensorData contains the contents and metadata for tabular data.
+type SensorData struct {
+	Metadata SensorMetadata
+	SDStruct map[string]interface{}
+	SDBinary []byte
+}
+
+// DataType specifies the type of data uploaded.
+type DataType int32
+
+// DataType constants define the possible DataType options.
+const (
+	DataTypeUnspecified DataType = iota
+	DataTypeBinarySensor
+	DataTypeTabularSensor
+	DataTypeFile
+)
+
+// MimeType specifies the format of a file being uploaded.
+type MimeType int32
+
+// MimeType constants define the possible MimeType options.
+const (
+	MimeTypeUnspecified MimeType = iota
+	MimeTypeJPEG
+	MimeTypePNG
+	MimeTypePCD
+)
+
+// UploadMetadata contains the metadata for binary (image + file) data.
+type UploadMetadata struct {
+	PartID           string
+	ComponentType    string
+	ComponentName    string
+	MethodName       string
+	Type             DataType
+	FileName         string
+	MethodParameters map[string]interface{}
+	FileExtension    string
+	Tags             []string
+}
+
+// FileData contains the contents of binary (image + file) data.
+type FileData struct {
+	Data []byte
+}
+
+// BinaryOptions represents optional parameters for the BinaryDataCaptureUpload method.
+type BinaryOptions struct {
+	Type             DataType
+	FileName         string
+	MethodParameters map[string]interface{}
+	Tags             []string
+	DataRequestTimes [2]time.Time
+}
```

Comment:
I'm not sure what the standard is in Go, but should these optional parameter structs sit all the way at the top with the shadow types? Personally, it makes it hard for me to read the code. Maybe someone else disagrees?

Reply:
I am looking at the `moveOptions` example used in arm and you can see it is defined [here](https://github.com/viamrobotics/rdk/blob/8c8f74d8809eaabccb8c0604f0167675c35c2dfa/components/arm/pb_helpers.go#L15) and we use it [here](https://github.com/viamrobotics/rdk/blob/8c8f74d8809eaabccb8c0604f0167675c35c2dfa/components/arm/client.go#L112)...what is different here is that they have an entire ph_helpers.go file dedicated to these helper structs and functions. This is definitely a cleaner way to go about it. what do u think? Or I can move the structs above the functions that are using them. @benjirewis is there a best practice here to follow?

Reply:
>  I'd be pretty surprised if there's an agreed-upon golang idiom here but @benjirewis if there is I'd love to hear it!

There is not a great agreed-upon Golang idiom! Apologies for jumping into the conversation late. I have seen more mature libraries with a large number of options structs put them in an `options` subpackage of the top level package. I think keeping them in the same file is fine and I have no strong opinion on whether the option struct is close to its associated function or at the top of file. Perhaps just remaining consistent in billing, app and data files is best.